internal/transport/transport_test.go

```diff
@@ -2746,3 +2746,100 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
 		t.Errorf("Context timed out")
 	}
 }
+
+// Test that in the event of a graceful client transport shutdown
+// , i.e., clientTransport.Close(), if the conn hung for
+// LoopyWriterTimeout, client should still be close itself and should
+// not wait for long.
+func (s) TestClientCloseReturnsEarlyWhenGoAwayWriteHangs(t *testing.T) {
+	// Override goAwayLoopyWriterTimeout to 0 so that we always
+	// time out while writing GOAWAY on client.Close(). This is
+	// equivalent to network hang scenario when client is
+	// failing to write GOAWAY frame.
+	origGoAwayLoopyTimeout := goAwayLoopyWriterTimeout
+	goAwayLoopyWriterTimeout = 0
+	defer func() {
+		goAwayLoopyWriterTimeout = origGoAwayLoopyTimeout
+	}()
+
+	// Create a server.
+	lis, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("Error while listening: %v", err)
+	}
+	defer lis.Close()
+	// serverGreetingDone is used to notify when server is done greeting the client.
+	serverGreetingDone := make(chan struct{})
+	// errorCh verifies that desired GOAWAY not received by server
+	errorCh := make(chan error, 1)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	// Launch the server.
+	go func() {
```

Comment:
Why do we need this custom server implementation?

What is it important for this server to notify when it is done reading the client preface?

Could you please add comments to describe why you are doing whatever you are doing here?